TypeCobol/Compiler/Sql/Scanner/SqlScanner.cs

```diff
@@ -50,46 +71,228 @@ public override Token GetTokenStartingFrom(int startIndex)
-                    case TokenType.SQL_SELECT:
-                    case TokenType.SQL_ALL:
-                    case TokenType.SQL_DISTINCT:
-                    case TokenType.UserDefinedWord:
-                    case TokenType.SQL_ROLLBACK:
-                    case TokenType.SQL_TO:
-                    case TokenType.SQL_SAVEPOINT:
-                    case TokenType.SQL_FROM:
-                    case TokenType.SQL_AS:
-                    case TokenType.SQL_TABLE:
-                    case TokenType.SQL_DELETE:
-                    case TokenType.SQL_IMMEDIATE:
-                    case TokenType.SQL_DROP:
-                    case TokenType.SQL_RESTRICT:
-                    case TokenType.SQL_WHEN:
-                    case TokenType.SQL_TRUNCATE:
-
-                        return new Token(tokenType, startIndex, currentIndex - 1, tokensLine);
-                    default:
-                        //Unrecognized keyword (for now) return as ExecStatementText
-                        return new Token(TokenType.ExecStatementText, startIndex, currentIndex - 1, tokensLine);
+                    //Out of range
+                    this.tokensLine.AddDiagnostic(MessageCode.SyntaxErrorInParser, token, "Decimal is too big: " + literalValue.Number);
                 }
             }
-            //Consume all sql-keyword incompatible chars
-            for (; currentIndex <= lastIndex && !IsSqlKeywordPart(line[currentIndex]); currentIndex++){}
-            return new Token(TokenType.ExecStatementText, startIndex, currentIndex - 1, tokensLine);
+
+            void CheckFloatingPointLiteral()
+            {
+                Debug.Assert(token.LiteralValue is FloatingPointLiteralTokenValue);
+                var literalValue = (FloatingPointLiteralTokenValue)token.LiteralValue;
+
+                //Check exponent and mantissa length
+                var exponent = literalValue.Exponent;
+                int exponentLength = exponent.Number.ToString().Length - (exponent.HasSign ? 1 : 0);
+                var mantissa = literalValue.Mantissa;
+                int mantissaLength = mantissa.IntegerValue.ToString().Length - (mantissa.HasSign ? 1 : 0);
+                if (exponentLength > 2 || mantissaLength > 17)
+                {
+                    //Try DecimalFloatingPointLiteral
+                    token.CorrectType(TokenType.SQL_DecimalFloatingPointLiteral);
+                    token.LiteralValue = new DecimalFloatingPointLiteralTokenValue(literalValue);
+                    CheckDecimalFloatingPointLiteral(exponentLength, mantissaLength);
+                }
+            }
+
+            void CheckDecimalFloatingPointLiteral(int exponentLength, int mantissaLength)
+            {
+                Debug.Assert(token.LiteralValue is DecimalFloatingPointLiteralTokenValue);
+                if (exponentLength > 4)
+                {
+                    tokensLine.AddDiagnostic(MessageCode.InvalidExponentInFloatingPointLiteral, token);
```

Comment:
We have to use another message because this one is Cobol specific. Please create a new error message with associated message code (csv + enum).
`The number of digits in the exponent of a decimal floating-point constant must not exceed 4.`

Reply:
Now that you created new messages, you have to use them so please replace `MessageCode.InvalidExponentInFloatingPointLiteral` with `MessageCode.InvalidExponentInDecimalFloatingPointLiteral`.